Slice up the development environment setup (#437)
zmkfirmware/zmk: docs/docs/development/setup/docker-vscode.md

---
title: Docker+VSCode Setup
sidebar_label: Docker+VSCode
---

:::danger The Docker+VSCode environment will **NOT** run on arm CPUs. You must use the Native environment if using an arm CPU.
:::

This setup leverages the same [image which is used by the GitHub action](https://github.com/zmkfirmware/zephyr-west-action) for local development. Beyond the benefits of [dev/prod parity](https://12factor.net/dev-prod-parity), this approach is also the easiest to set up. No toolchain or dependencies are necessary when using Docker; the container image you'll be using already has the toolchain installed and set up to use.

1. Install [Docker Desktop](https://www.docker.com/products/docker-desktop) for your operating system.
2. Install [VS Code](https://code.visualstudio.com/)
3. Install the [Remote - Containers extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers)

:::info

> Once the above is setup you're ready to setup the ZMK sources as defined [here](/docs/development/setup/zmk)

Reviewer: See above.

---
title: Docker+VS Code Setup
sidebar_label: Docker
---

:::danger

> The Docker environment will **NOT** run on arm CPUs like the Raspberry Pi. You must use the native environment if using an arm CPU.

Reviewer 2: Probably worth calling out apple silicon here as well

```suggestion
The Docker environment will **NOT** run on arm CPUs like the Raspberry Pi or Apple Silicon. You must use the native environment if using an arm CPU.
```

:::

:::note Windows Users

> Please note the zmk builds can run slower with Docker on Windows if you don't use the WSL2 filesystem to store files. Build times can take 3-5 minutes on slower hardware without using the WSL2 filesystem. If you run into performance problems you can checkout the zmk sources inside a WSL2 environment and use `code .` inside the WSL2 environment to open the sources. This can make builds run as fast as 20s.

Reviewer 2: If the path of least resistance is installing Docker for Windows and letting that work the WSL2 magic behind the scenes, maybe we ought to recommend the happiest path upfront. Can also link to a section at the end that talks about the finer points of WSL and IO perf.

Author: I think we can leave this note but move it after the setup steps.

 @innovaker was the one driving this add initially and I don't think it detracts from the rest of the information on the page.

> This setup leverages the same [image which is used by the GitHub action](https://github.com/zmkfirmware/zephyr-west-action) for local development. Beyond the benefits of [dev/prod parity](https://12factor.net/dev-prod-parity), this approach is also the easiest to set up. No toolchain or dependencies are necessary when using Docker; the container image you'll be using already has the toolchain installed and set up to use.

Author: It's there to be clear that the container used for local dev is the same as used by the github build system.

Basically a signal that if you want to be guaranteed what you do locally (dev) is the same as on the upstream build stuff (prod) you should be using the docker container for builds.

This should stand IMO for folks (like me) who want to use the same environment locally as is used by the build system to ensure there aren't surprises.

> This approach will also need the [Remote - WSL](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-wsl) extension installed in VS Code as well.

Reviewer: ```suggestion
This approach will also need the [Remote - WSL](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-wsl) extension installed in VS Code.
```
"also" and "as well" are redundant. Only need one or the other.

3. Install the [Remote - Containers extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers)

:::note Windows Users

> Please note the zmk builds can run slower with Docker on Windows if you don't use the WSL2 filesystem to store files. Build times can take 3-5 minutes on slower hardware without using the WSL2 filesystem. If you run into performance problems you can checkout the zmk sources inside a WSL2 environment and use `code .` inside the WSL2 environment to open the sources. This can make builds run as fast as 20s.

Reviewer 3: ```suggestion
Please note the ZMK builds run slower (up to 3-5 minutes on slower hardware) with Docker on Windows if you don't use the WSL2 filesystem to store files. If you run into performance problems you can checkout the ZMK sources inside a WSL2 environment and run `code .` to open the sources. This can make builds run at near-native speed.
```
